### ISSUE TITLE
sf::Image::saveToFile() fails when user has special characters in his username

### ISSUE THREAD

**Issue by GuiTeK** [SFML/SFML#647]
When a user has special characters in his username (e.g. "Sébastien", notice the "é"), and he chooses to save an image under a path which contains the username (e.g. C:\Users\%USERNAME%\Desktop under Windows), sf::Image::saveToFile() fails with the error: "Failed to save image FILENAME" (file ImageLoader.cpp, line 261, SFML 2.1).

It would be good if sf::Image::saveToFile() was overloaded in order to accept an std::wstring instead of an std::string.

**Comment by binary1248** [comment]
Until SFML is the one to provide the low-level IO that its dependency libraries depend on, there is no real way to solve this.

Once SFML has UTF-8 file stream support, we can simply pass those streams to the libraries in question and they won't have to make use of their own low-level IO which might not support UTF-8 or UTF-16 filenames.

**Comment by Quintus** [comment]
I just stumbled over this thread and thought I’d give you a quick note on what I experienced with regard to cross-platform Unicode file pathes:

* On Windows, MinGW and VisualStudio have different constructors for the `std::ifstream`, etc. classes. VisualStudio has a nonstandard constructor that takes a `wchar_t` that it assumes to be in UTF-16LE; this constructor doesn’t exist on MinGW so there’s **no C++ish possibility in MinGW** to obtain a handle to a Unicode file path, however:
* The [`_wfopen()`](https://msdn.microsoft.com/en-us/library/yeby3zcb.aspx) C function allows to obtain a `FILE*` pointer to a filename encoded in UTF16LE in both MinGW and VS, so once you have that, the interface is equal (C functions).
* GCC provides a nonstandard interface to [construct a C++ I/O stream from a `FILE*`](https://gcc.gnu.org/onlinedocs/gcc-4.9.2/libstdc++/api/a00054.html), which surprisingly also works on Windows.
* On Linux, the file system encoding is dependant on the locale, which today is mostly UTF-8. However, by adjusting the `LANG` environment variable, even a single user could have files with differently encoded filenames.
* This does not hurt, because on Linux (and probably other UNIX systems), the usual `char`-based C and C++ interfaces assume the regular `char`-based strings to be in the locale encoding, i.e. mostly UTF-8, and work as expected even with multibyte characters.
* I have no idea on OS X as I don’t use it.

On top of this information I’ve earlier implemented a library that abstracts the problem away under an interface that always takes UTF-8 strings on any (= Linux, MinGW Windows, VS Windows) platform (the library does other things also, but that’s not important here). You might want to take a look at it for obtaining an overview of the problem:

* [Header file](https://github.com/Quintus/pathie-cpp/blob/devel/include/pathie_ofstream.hpp)
* [Implementation](https://github.com/Quintus/pathie-cpp/blob/devel/src/pathie_ofstream.cpp)
* [Documentation](https://quintus.github.io/pathie-cpp/)

I’m not familiar with the SFML codebase (I’m just learning SFML), otherwise I’d try my hands at a PR. But before I haven’t really learned the library, I thought I’d rather not do it and instead describe how I coped with Unicode pathnames :-)

Disclaimer: I haven’t tested my findings regarding Visual Studio, I only use MinGW for Windows (by crosscompilation), but that’s what I got from several intense web searches.

Greetings
Marvin

**Comment by eXpl0it3r** [comment]
> Until SFML is the one to provide the low-level IO that its dependency libraries depend on, there is no real way to solve this.

Since this is marked as 2.4, should someone open an issue for low-level IO implementations?

**Comment by binary1248** [comment]
I guess that would make sense...

**Comment by eXpl0it3r** [comment]
Since this will requires quite a bit more work, I'll remove the 2.4 milestone.

**Comment by ChrisThrasher** [comment]
#1964 is starting to address this problem with C++17’s Filesystem library. This lets us seamlessly support any OS’s paths including Windows and its use of wide strings and means we don’t need to add `std::wstring` overloads anywhere which is great. However, we still have the same problem from years ago that some of SFML’s dependencies use `const char*` for paths. Where those 3rd party APIs are not needed, we can fully support non-ASCII paths but where those are APIs _are_ needed then I’m not sure if we can support non-ASCII paths.

**Comment by Pixel-Tony** [comment]
@ChrisThrasher since support for fs::path was added, this one can be closed?

**Comment by FRex** [comment]
@Pixel-Tony I do not think so.

Image still uses the `char * ` versions of save functions in here: https://github.com/SFML/SFML/blob/master/src/SFML/Graphics/Image.cpp#L316C24-L316C61

This simple test program ilustrates that on Windows that is not enough:

```cpp
#include <filesystem>

int main(void)
{
    const std::filesystem::path fname1 = L"\x105"; // a with ogonek
    const std::filesystem::path fname2 = L"\x119"; // e with ogonek
    fopen(fname1.string().c_str(), "wb");
    _wfopen(fname2.wstring().c_str(), L"wb");
}
```

With MSVC this program crashes, if I comment out the `fopen` it creates the `ę` file. With GCC the `ę` file is created, but the `ą` is `Ä…` instead.

I used the `\x` values is to not rely on compiler to pass through UTF-8 literals from source to end program. I used ą/ę instead of á/é or similar, to be sure I won't get lucky and characters won't be in my PC's codepage, which would make `char*` version work. Some multi-language name (including emojis which are non-BMP, so they take 2 whcars) filenames can be used to be 100% sure when testing such feature later, including emojis (to hit some non-BMP characters which take 2 whcars). In my own programs I usually put in English, Russian, Chinese and Emoji characters all in single filename when testing unicode filenames.

Note that `stb_image_write` as of 2015 has added callbacks (this issue and our comments above saying it doesn't support non-ASCII filenames on Windows is from 2014, so before that) so we can save to any arbitrary data sink: https://github.com/nothings/stb/blob/master/stb_image_write.h#L58

**Comment by ChrisThrasher** [comment]
> Note that `stb_image_write` as of 2015 has added callbacks (this issue and our comments above saying it doesn't support non-ASCII filenames on Windows is from 2014, so before that) so we can save to any arbitrary data sink: https://github.com/nothings/stb/blob/master/stb_image_write.h#L58

Is the idea here that we create the file using the `std::filesystem::path` thus preserving the special encoding, then provide a function that stb_image uses to write to that file we created ourselves? I suppose that's doable. I'm glad to know it exists.

**Comment by FRex** [comment]
Yes, apparently C++17's `std::ofstream` allows a `std::filesystem::path` argument. I just tried and both with MSVC and GCC I have the below program worked:

```cpp
#include <filesystem>
#include <iostream>
#include <fstream>

int main(void)
{
    // "日本語 русский язык ABC test🤔👀🐌.txt" as UTF16LE bytes
    const unsigned char fname_utf16le[] = {
        229, 101, 44, 103, 158, 138, 32, 0, 64, 4, 67, 4, 65, 4, 65, 4,
        58, 4, 56, 4, 57, 4, 32, 0, 79, 4, 55, 4, 75, 4, 58, 4, 32, 0,
        65, 0, 66, 0, 67, 0, 32, 0, 116, 0, 101, 0, 115, 0, 116, 0, 62, 216,
        20, 221, 61, 216, 64, 220, 61, 216, 12, 220, 46, 0, 116, 0, 120, 0, 116, 0,
        0, 0
    };
    const wchar_t * fname = (const wchar_t*)fname_utf16le;
    std::filesystem::path filename{fname};
    std::ofstream file{filename};
    std::cout << file.good() << std::endl;
}
```

The filename is `日本語 русский язык ABC test🤔👀🐌.txt`. It has CJK/Japanese, Russian/cyrillic, English/Latin1/ASCII and a few non-BMP (codepoints above `0xffff`) emotes. This makes sure multiple languages in one filename, uncommon (for us) languages, and non-BMP codepoints (the ones that take 2 whcars) work all properly. Just something like `español áíúéóü` is not ASCII but it's Latin1, and for (some?) American user with American Latin1-ish codepage it'd work even as `char`.

Here's a link to a Unicode inspector website for that filename: https://frex.github.io/unicode.html?text=%E6%97%A5%E6%9C%AC%E8%AA%9E%20%D1%80%D1%83%D1%81%D1%81%D0%BA%D0%B8%D0%B9%20%D1%8F%D0%B7%D1%8B%D0%BA%20ABC%20test%F0%9F%A4%94%F0%9F%91%80%F0%9F%90%8C.txt

**Comment by ChrisThrasher** [comment]
In that case I'm definitely open to using those alternative STB functions. When I transitioned SFML 3 to `std::filesystem::path` I knew it wouldn't make a big short term benefit given the fact that we use C APIs which just use `const char*` for file names. However, I also knew it would leave the door open for future internal refactoring to make use of that extra info. So this is definitely in line with what I had in mind.

**Comment by ChrisThrasher** [comment]
It only took a decade but this is fixed with #3403.
